catalog/spatial/geocoding/spatial-geocoding-plugin/src/main/java/org/codice/ddf/spatial/geocoding/plugin/GeoCoderPlugin.java

```diff
@@ -0,0 +1,121 @@
+/**
+ * Copyright (c) Codice Foundation
+ * <p/>
+ * This is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser
+ * General Public License as published by the Free Software Foundation, either version 3 of the
+ * License, or any later version.
+ * <p/>
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.spatial.geocoding.plugin;
+
+import java.io.Serializable;
+import java.util.Map;
+import java.util.Optional;
+
+import org.codice.ddf.spatial.geocoder.GeoCoder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import ddf.catalog.data.Metacard;
+import ddf.catalog.data.impl.AttributeImpl;
+import ddf.catalog.data.types.Location;
+import ddf.catalog.operation.CreateRequest;
+import ddf.catalog.operation.DeleteRequest;
+import ddf.catalog.operation.UpdateRequest;
+import ddf.catalog.plugin.PluginExecutionException;
+import ddf.catalog.plugin.PreIngestPlugin;
+import ddf.catalog.plugin.StopProcessingException;
+import ddf.catalog.util.impl.ServiceSelector;
+
+/**
+ * The {@code GeoCoderPlugin} is responsible for adding a {@link Location#COUNTRY_CODE} in ISO 3166-1 format
+ * to {@code Metacard}s that have the {@link Metacard#GEOGRAPHY} attribute.
+ */
+public class GeoCoderPlugin implements PreIngestPlugin {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(GeoCoderPlugin.class);
+
+    private static ServiceSelector<GeoCoder> geoCoderFactory;
+
+    private int radiusInKm = 10;
+
+    public GeoCoderPlugin(ServiceSelector<GeoCoder> geoCoderFactory) {
+        if (geoCoderFactory == null) {
+            throw new IllegalArgumentException(
+                    "GeoCoderPlugin(): constructor argument 'geoCoderFactory' may not be null.");
+        }
+
+        this.geoCoderFactory = geoCoderFactory;
+    }
+
+    @Override
+    public CreateRequest process(CreateRequest input)
+            throws PluginExecutionException, StopProcessingException {
+        if (input.getMetacards() == null) {
+            return input;
+        }
+
+        for (Metacard metacard : input.getMetacards()) {
+            Optional<String> wktLocation = Optional.ofNullable(metacard.getLocation());
+
+            if (wktLocation.isPresent()) {
+                if (countryCodeNeeded(metacard)) {
+                    GeoCoder geocoder = geoCoderFactory.getService();
```

Comment:
what happens if a geocoder isn't available?  For instance, if one of the gazetteer features isn't installed?